File: server/internal/model/feed.go
package model

import "time"

// Topic 是用户关心的话题（关键词/事项）。
// 同一个用户下 (user_id, name) 唯一；weight 用作匹配时的权重放大。
type Topic struct {
	ID        int64     `json:"id"`
	UserID    int64     `json:"user_id"`
	Name      string    `json:"name"`
	Weight    float64   `json:"weight"`
	CreatedAt time.Time `json:"created_at"`
}

// NewsSource 是一个新闻数据源（RSS / atom feed）。
type NewsSource struct {
	ID          int64      `json:"id"`
	Name        string     `json:"name"`
	URL         string     `json:"url"`
	Kind        string     `json:"kind"`   // rss
	Region      string     `json:"region"` // cn / intl_zh / intl_en
	Lang        string     `json:"lang"`   // zh / en
	Category    string     `json:"category"`    // tech / ai / finance / intl / sports / culture / dev …
	Description string     `json:"description"` // 1-2 句源简介，给 LLM 推荐用
	Enabled     bool       `json:"enabled"`
	LastFetchAt *time.Time `json:"last_fetch_at,omitempty"`
	LastError   string     `json:"last_error,omitempty"`
	CreatedAt   time.Time  `json:"created_at"`
}

// NewsEvent 是一条入库的新闻事件。
type NewsEvent struct {
	ID          int64      `json:"id"`
	SourceID    int64      `json:"source_id"`
	URL         string     `json:"url"`
	Title       string     `json:"title"`
	Summary     string     `json:"summary,omitempty"`
	Content     string     `json:"content,omitempty"`
	Lang        string     `json:"lang"`
	PublishedAt *time.Time `json:"published_at,omitempty"`
	FetchedAt   time.Time  `json:"fetched_at"`
	Extracted   bool       `json:"extracted"`
}

// Entity 是关系图谱里的节点：人物 / 机构 / 地点 / 概念 / 事件。
type Entity struct {
	ID          int64     `json:"id"`
	Type        string    `json:"type"`
	Name        string    `json:"name"`
	Weight      float64   `json:"weight"`
	FirstSeenAt time.Time `json:"first_seen_at"`
	LastSeenAt  time.Time `json:"last_seen_at"`
}

// EventEntity 把一条事件和它涉及的实体连起来，salience 用作显著度。
type EventEntity struct {
	EventID  int64   `json:"event_id"`
	EntityID int64   `json:"entity_id"`
	Salience float64 `json:"salience"`
}

// EntityRelation 是图谱里的边：src→dst，label 描述关系语义。
// 同 (src, dst, label) 多次出现会累加 weight + 刷新 last_seen_at。
type EntityRelation struct {
	ID         int64     `json:"id"`
	SrcID      int64     `json:"src_id"`
	DstID      int64     `json:"dst_id"`
	Label      string    `json:"label"`
	Weight     float64   `json:"weight"`
	LastSeenAt time.Time `json:"last_seen_at"`
}

// UserEventSub 是「这个用户因为某些 topic 命中了这条事件」的记录。
type UserEventSub struct {
	UserID        int64     `json:"user_id"`
	EventID       int64     `json:"event_id"`
	Relevance     float64   `json:"relevance"`
	MatchedTopics []int64   `json:"matched_topics"`
	CreatedAt     time.Time `json:"created_at"`
}

// FeedStatus 是事件流 worker 的运行态快照。
type FeedStatus struct {
	Running       bool       `json:"running"`
	LastFetchAt   *time.Time `json:"last_fetch_at,omitempty"`
	LastPruneAt   *time.Time `json:"last_prune_at,omitempty"`
	SourcesTotal  int        `json:"sources_total"`
	SourcesActive int        `json:"sources_active"`
	Events24h     int        `json:"events_24h"`
	EntitiesTotal int        `json:"entities_total"`
	RelationsTotal int       `json:"relations_total"`
	LastError     string     `json:"last_error,omitempty"`
}

File: app/lib/models/feed.dart
/// 事件流图谱相关的数据模型。
///
/// 后端 JSON 都使用 snake_case，前端用 camelCase 保留 Dart 风格。
class FeedStatus {
  FeedStatus({
    required this.running,
    required this.sourcesTotal,
    required this.sourcesActive,
    required this.events24h,
    required this.entitiesTotal,
    required this.relationsTotal,
    this.lastFetchAt,
    this.lastPruneAt,
    this.lastError = '',
  });

  final bool running;
  final int sourcesTotal;
  final int sourcesActive;
  final int events24h;
  final int entitiesTotal;
  final int relationsTotal;
  final DateTime? lastFetchAt;
  final DateTime? lastPruneAt;
  final String lastError;

  bool get healthy => running && lastError.isEmpty;

  factory FeedStatus.fromJson(Map<String, dynamic> j) => FeedStatus(
        running: j['running'] == true,
        sourcesTotal: (j['sources_total'] ?? 0) as int,
        sourcesActive: (j['sources_active'] ?? 0) as int,
        events24h: (j['events_24h'] ?? 0) as int,
        entitiesTotal: (j['entities_total'] ?? 0) as int,
        relationsTotal: (j['relations_total'] ?? 0) as int,
        lastFetchAt: _parseTime(j['last_fetch_at']),
        lastPruneAt: _parseTime(j['last_prune_at']),
        lastError: (j['last_error'] as String?) ?? '',
      );
}

class Topic {
  Topic({
    required this.id,
    required this.name,
    required this.weight,
    required this.createdAt,
  });

  final int id;
  final String name;
  final double weight;
  final DateTime createdAt;

  factory Topic.fromJson(Map<String, dynamic> j) => Topic(
        id: (j['id'] as num).toInt(),
        name: j['name'] as String,
        weight: (j['weight'] as num?)?.toDouble() ?? 1.0,
        createdAt: _parseTime(j['created_at']) ?? DateTime.now(),
      );
}

class FeedEvent {
  FeedEvent({
    required this.id,
    required this.url,
    required this.title,
    required this.summary,
    required this.lang,
    required this.fetchedAt,
    required this.sourceId,
    required this.sourceName,
    required this.relevance,
    required this.matchedTopics,
    this.publishedAt,
  });

  final int id;
  final String url;
  final String title;
  final String summary;
  final String lang;
  final DateTime fetchedAt;
  final DateTime? publishedAt;
  final int sourceId;
  final String sourceName;
  final double relevance;
  final List<int> matchedTopics;

  factory FeedEvent.fromJson(Map<String, dynamic> j) => FeedEvent(
        id: (j['id'] as num).toInt(),
        url: (j['url'] as String?) ?? '',
        title: (j['title'] as String?) ?? '',
        summary: (j['summary'] as String?) ?? '',
        lang: (j['lang'] as String?) ?? 'zh',
        fetchedAt: _parseTime(j['fetched_at']) ?? DateTime.now(),
        publishedAt: _parseTime(j['published_at']),
        sourceId: (j['source_id'] as num?)?.toInt() ?? 0,
        sourceName: (j['source_name'] as String?) ?? '',
        relevance: (j['relevance'] as num?)?.toDouble() ?? 0,
        matchedTopics: ((j['matched_topics'] as List?) ?? const [])
            .map((e) => (e as num).toInt())
            .toList(),
      );
}

class GraphNode {
  GraphNode({
    required this.id,
    required this.type,
    required this.name,
    required this.weight,
    required this.lastSeenAt,
  });

  final int id;
  final String type;
  final String name;
  final double weight;
  final DateTime lastSeenAt;

  factory GraphNode.fromJson(Map<String, dynamic> j) => GraphNode(
        id: (j['id'] as num).toInt(),
        type: (j['type'] as String?) ?? 'concept',
        name: (j['name'] as String?) ?? '',
        weight: (j['weight'] as num?)?.toDouble() ?? 1.0,
        lastSeenAt: _parseTime(j['last_seen_at']) ?? DateTime.now(),
      );
}

class GraphEdge {
  GraphEdge({
    required this.id,
    required this.srcId,
    required this.dstId,
    required this.label,
    required this.weight,
  });

  final int id;
  final int srcId;
  final int dstId;
  final String label;
  final double weight;

  factory GraphEdge.fromJson(Map<String, dynamic> j) => GraphEdge(
        id: (j['id'] as num).toInt(),
        srcId: (j['src_id'] as num).toInt(),
        dstId: (j['dst_id'] as num).toInt(),
        label: (j['label'] as String?) ?? 'related',
        weight: (j['weight'] as num?)?.toDouble() ?? 1.0,
      );
}

class FeedGraph {
  FeedGraph({required this.nodes, required this.edges});

  final List<GraphNode> nodes;
  final List<GraphEdge> edges;

  bool get isEmpty => nodes.isEmpty;

  factory FeedGraph.fromJson(Map<String, dynamic> j) => FeedGraph(
        nodes: ((j['nodes'] as List?) ?? const [])
            .map((e) => GraphNode.fromJson(e as Map<String, dynamic>))
            .toList(),
        edges: ((j['edges'] as List?) ?? const [])
            .map((e) => GraphEdge.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

DateTime? _parseTime(dynamic v) {
  if (v == null || v == '') return null;
  if (v is String) return DateTime.tryParse(v);
  return null;
}

/// 新闻源（给「源库管理」用）。
class NewsSource {
  NewsSource({
    required this.id,
    required this.name,
    required this.url,
    required this.category,
    required this.description,
    required this.enabled,
    required this.region,
    required this.lang,
  });

  final int id;
  final String name;
  final String url;
  final String category;
  final String description;
  final bool enabled;
  final String region;
  final String lang;

  factory NewsSource.fromJson(Map<String, dynamic> j) => NewsSource(
        id: (j['id'] as num).toInt(),
        name: (j['name'] as String?) ?? '',
        url: (j['url'] as String?) ?? '',
        category: (j['category'] as String?) ?? 'general',
        description: (j['description'] as String?) ?? '',
        enabled: j['enabled'] == true,
        region: (j['region'] as String?) ?? 'cn',
        lang: (j['lang'] as String?) ?? 'zh',
      );
}

/// AI 情报简报元数据（列表项用，正文走 HTML 接口）
class Briefing {
  Briefing({
    required this.id,
    required this.window,
    required this.title,
    required this.summary,
    required this.model,
    required this.eventCount,
    required this.clusterCount,
    required this.generatedAt,
  });

  final int id;
  final String window;
  final String title;
  final String summary;
  final String model;
  final int eventCount;
  final int clusterCount;
  final DateTime generatedAt;

  factory Briefing.fromJson(Map<String, dynamic> j) => Briefing(
        id: (j['id'] as num).toInt(),
        window: (j['window'] as String?) ?? '1h',
        title: (j['title'] as String?) ?? '',
        summary: (j['summary'] as String?) ?? '',
        model: (j['model'] as String?) ?? '',
        eventCount: (j['event_count'] as num?)?.toInt() ?? 0,
        clusterCount: (j['cluster_count'] as num?)?.toInt() ?? 0,
        generatedAt: _parseTime(j['generated_at']) ?? DateTime.now(),
      );
}

/// LLM 智能选源结果。
class RecommendResult {
  RecommendResult({required this.newlyEnabled, required this.reason});

  final List<NewsSource> newlyEnabled;
  final String reason;

  factory RecommendResult.fromJson(Map<String, dynamic> j) => RecommendResult(
        newlyEnabled: ((j['newly_enabled'] as List?) ?? const [])
            .map((e) => NewsSource.fromJson(Map<String, dynamic>.from(e)))
            .toList(),
        reason: (j['reason'] as String?) ?? '',
      );
}
